### PR TITLE
Only allow primitive JSON for transfer between worker and main worker

### LINKED ISSUE #96: Improve type definition for Data
We could try to use following types to improve the type definition of `Data`

[TS Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAUgygeQHIAUBOBLAth4GBu0AvFAHYCuWARhGlAD5RUD2zANhAIakNQDOwTKQDmvCmzYBuAFChIsREgBqnNuWILk6bLgLRG8ZAEE0aTiF6GkCKgCsIAY2Ay50KzftOoJAN5QA2gDWAFz8ghgiALqhVipqEJJQAL4u4G6KJmYWJJnmADyxquoAfDLSDsykAlAAJpzAnACMMYpx6t5QPinlldV1DQBMLchtGgDkwBACYzIVVcC19ZwAzMPKRRoALAOzvQv9nJtrox3iUtLlbJx8fFAAIiCknDgOACoAFmhcNSisbHnSKD3JZQCAAD0mpBqt0K8Q6sPUABpAVAAEpTMC9aDgyHQzTrOEkBEQaTFToooFgchUNgYBygsGOciTAAUB1CdyWAEpQuhmDg+BA8ui+JiqhAyT4KUCZV9gOQ0Dw+QKIAA6L58diEFl+TihRoADkRTH1A2SUGu+JOlpFYsFXJkMuS0iSF1IEAA7vdHs86R8vpwfn8AU7daEKNRaMlkaGLeHKDQ0MaqPHI3RXcUWQ6gA)

So we could e.g. write

```ts
export class DynamicThreadPool<
  Data extends JSONValue = JSONValue,
  Response extends JSONValue = JSONValue
> extends FixedThreadPool<Data, Response>
```

### PR DESCRIPTION
**Pull request by Shinigami92**
closes #96

### CODE REVIEW

**Comment by Shinigami92** [comment]
@pioardi 
I think this theoretically could be a _breaking_ change for users that already use typescript :thinking:  
So I should note add a note to the changelog

I will make a test locally a bit later and report how it feels and how much this improve the usability  

You can also try it by using `yarn link` (or its npm equivalent) and create a project and the use `yarn link "poolifier"` to override your dependency to use your local copy/repo of poolifier

**Comment by pioardi** [comment]
mmmmm so I need to think about those changes( regardless the breaking change).
We really want this restriction I mean we could just advice people that worker_thread and/or cluster processes does not support some types of input mmmmmmmm.....need to think about that

**Comment by Shinigami92** [comment]
> mmmmm so I need to think about those changes( regardless the breaking change).
> We really want this restriction I mean we could just advice people that worker_thread and/or cluster processes does not support some types of input mmmmmmmm.....need to think about that

When I first used poolifier with a colleague at adSoul, we tried to move a non-serializable object between worker and main worker.  
Of course it didn't work, but by then it was already "too late".  
We just thought wrong and didn't look in the node documentation. #RTFM
If we had direct typing, it would have been noticed immediately.

That's the reason why I want to type it :slightly_smiling_face:

File: src/utility-types.ts
Comment (pioardi) [inline src/utility-types.ts:3]: we are sure that there are no other JSON primitives that we want to list here?

**Comment by pioardi** [comment]
> > mmmmm so I need to think about those changes( regardless the breaking change).
> > We really want this restriction I mean we could just advice people that worker_thread and/or cluster processes does not support some types of input mmmmmmmm.....need to think about that
> 
> When I first used poolifier with a colleague at adSoul, we tried to move a non-serializable object between worker and main worker.
> Of course it didn't work, but by then it was already "too late".
> We just thought wrong and didn't look in the node documentation. #RTFM
> If we had direct typing, it would have been noticed immediately.
> 
> That's the reason why I want to type it 🙂

Make sense make sense, my concern is more to avoid any too much restrictive type( I also added a comment).

File: src/utility-types.ts
Comment (Shinigami92) [reply src/utility-types.ts:3]: Which are you thinking of?

Comment (pioardi) [reply src/utility-types.ts:3]: Nothing specific but you know types can be more and more and if we miss any here we will cause bugs.
This is my concern

Comment (Shinigami92) [reply src/utility-types.ts:3]: There are only a limited number of primitive types
The latest (`bigint`) is so new that few scientists use it for big calculations

https://developer.mozilla.org/en-US/docs/Glossary/Primitive

> There are 7 primitive data types: string, number, bigint, boolean, undefined, symbol, and null.

- `string`, `number`, `boolean` and `null` are used
- `undefined` is covered by the indexed object `JSONObject`
- `symbol` cannot be serialized as JSON values
- `bigint` could be added but also this is _currently_ [not supported](https://golb.hplar.ch/2019/01/js-bigint-json.html) by JSON :warning:  

... so indeed we have every JSON primitiv covered

Comment (pioardi) [reply src/utility-types.ts:3]: mmmmmmmmmmmmmmm @jerome-benoit  what do you think about this?
@Shinigami92 I think is really useful to define this and document it in JSDoc, but we should not apply that to our interface but on the concrete implementations, this is because we don't know if in future some other pool will support any type

Comment (Shinigami92) [reply src/utility-types.ts:3]: That indeed makes sense! So I will remove it from the abstract Pool and Worker.  
I will try if `unknown` can be applied for the abstraction, so we don't use the evil `any` type

Comment (pioardi) [reply src/utility-types.ts:3]: Great, I guess we can proceed I think this is beneficial for users

**Review by pioardi — DISMISSED** [review]
Thanks, this help end users to avoid using inputs that are not supported by node js engine

**Comment by Shinigami92** [comment]
> Thanks, this help end users to avoid using inputs that are not supported by node js engine

@pioardi Just a thing for approvals: you should approve a PR when it's out of draft && the author requested your review
Otherwise when pushing additional commits, the approval gets stale and I need to re-request your approval :disappointed:  

---

Now you can approve so I can merge :smiley:

File: CHANGELOG.md
Comment (pioardi) [inline CHANGELOG.md:31]: I would add that this is due to a nodejs limitation and does not depends by this library.

**Review by pioardi — APPROVED** [review]
Good ti go

File: src/utility-types.ts
Comment (jerome-benoit) [reply src/utility-types.ts:3]: > That indeed makes sense! So I will remove it from the abstract Pool and Worker.
> I will try if `unknown` can be applied for the abstraction, so we don't use the evil `any` type
> 
> Edit: done

You can also use Record<string, unknown> to ensure a key/value structure is required.

File: src/utility-types.ts
Comment (jerome-benoit) [inline src/utility-types.ts:6]: The type shortcut is better: Record<string, JSONValue>

Comment (Shinigami92) [reply src/utility-types.ts:6]: This is NOT possible, because with that you get a reference loop

**Pull request merged**
